hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -0,0 +1,370 @@
+    require(elementType.isNumeric)
+    val aoff = mb.newLocal[Long]
+    Code(
+      aoff := allocate(region, length),
+      stagedInitialize(aoff, length),
+      Region.setMemory(aoff + elementsOffset(length), length.toL * elementByteSize, 0.toByte),
+      aoff)
+  }
+
+  def anyMissing(mb: MethodBuilder, aoff: Code[Long]): Code[Boolean] =
+    if (elementType.required)
+      false
+    else {
+      val n = mb.newLocal[Long]
+      JoinPoint.CallCC[Code[Boolean]] { (jb, ret) =>
+        val loop = jb.joinPoint[Code[Long]](mb)
+        loop.define { ptr =>
+          (ptr < n).mux(
+            Region.loadInt(ptr).cne(0).mux(
+              ret(true),
+              loop(ptr + 4L)),
+            (Region.loadByte(ptr) >>>
+              (const(32) - (loadLength(aoff) | 31))).cne(0).mux(
+              ret(true),
+              ret(false)))
+        }
+        Code(
+          n := aoff + ((loadLength(aoff) >>> 5) * 4 + 4).toL,
+          loop(aoff + 4L))
+      }
+    }
+
+  def forEach(mb: MethodBuilder, region: Code[Region], aoff: Code[Long], body: Code[Long] => Code[Unit]): Code[Unit] = {
+    val i = mb.newLocal[Int]
+    val n = mb.newLocal[Int]
+    Code(
+      n := loadLength(aoff),
+      i := 0,
+      Code.whileLoop(i < n,
+        isElementDefined(aoff, i).mux(
+          body(loadElement(region, aoff, n, i)),
+          Code._empty
+        )))
+  }
+
+  override def unsafeOrdering(): UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering = {
+    require(this.isOfType(rightType))
+
+    val right = rightType.asInstanceOf[PCanonicalArray]
+    val eltOrd = elementType.unsafeOrdering(
+      right.elementType)
+
+    new UnsafeOrdering {
+      override def compare(r1: Region, o1: Long, r2: Region, o2: Long): Int = {
+        val length1 = loadLength(r1, o1)
+        val length2 = right.loadLength(r2, o2)
+
+        var i = 0
+        while (i < math.min(length1, length2)) {
+          val leftDefined = isElementDefined(r1, o1, i)
+          val rightDefined = right.isElementDefined(r2, o2, i)
+
+          if (leftDefined && rightDefined) {
+            val eOff1 = loadElement(r1, o1, length1, i)
+            val eOff2 = right.loadElement(r2, o2, length2, i)
+            val c = eltOrd.compare(r1, eOff1, r2, eOff2)
+            if (c != 0)
+              return c
+          } else if (leftDefined != rightDefined) {
+            val c = if (leftDefined) -1 else 1
+            return c
+          }
+          i += 1
+        }
+        Integer.compare(length1, length2)
+      }
+    }
+  }
+
+  def hasMissingValues(sourceOffset: Code[Long]): Code[Boolean] = {
+    if(elementType.required) {
+      return const(false)
+    }
+
+    Region.containsNonZeroBits(sourceOffset + lengthHeaderBytes, loadLength(sourceOffset).toL)
+  }
+
+  def checkedConvertFrom(mb: EmitMethodBuilder, r: Code[Region], sourceOffset: Code[Long], sourceType: PContainer, msg: String): Code[Long] = {
```

Comment:
Do we want methods that take a PContainer to be defined in the child class? If not, we can make it generic on :>PContainer

Reply:
sorry, don't totally understand what you mean. Can you clarify?

Reply:
I meant: should since this method is defined on PCanonicalArray, should it take a sourceType: PArray? This would also remove the need for the `assert(this.isOfType(sourceType))`